File: spark/src/main/scala/geotrellis/spark/io/LayerQuery.scala
package geotrellis.spark.io

import geotrellis.spark._
import geotrellis.util._

/**
  * Accumulation of [[LayerFilter]]s that will be asked to filter layer [[KeyBounds]]
  *
  * @tparam K  Type of key for the RDD being filtered
  * @tparam M  Type of metadata used for filtering
  */
class LayerQuery[K: Boundable, M: GetComponent[?, Bounds[K]]](
  filterChain: ( (M, List[KeyBounds[K]]) ) => (M, List[KeyBounds[K]]) = { x: (M, List[KeyBounds[K]]) => x }) {

  /**
    * @param metadata RasterMetaData of the layer being queried
    * @return A sequence of [[KeyBounds]] that cover the queried region
    */
  def apply(metadata: M): Seq[KeyBounds[K]] =
    metadata.getComponent[Bounds[K]] match {
      case keyBounds: KeyBounds[K] =>
        val (_, keyBoundsList) = filterChain((metadata, List(keyBounds)))
        keyBoundsList
      case EmptyBounds =>
        Seq()
    }

  // Allows us to treat Function1 as an instance of a Functor
  import scalaz.Scalaz._

  def where[F, T](exp: LayerFilter.Expression[F, T])(implicit filter: LayerFilter[K, F, T, M]): LayerQuery[K, M] = {
    new LayerQuery( {
      filterChain map { case (metadata, keyBoundsList) =>
        val filteredKeyBounds =
          for (keyBound <- keyBoundsList) yield {
            filter(metadata, keyBound, exp)
          }
        (metadata, filteredKeyBounds.flatten)
      }
    })
  }
}

/**
  * Wrapper for [[LayerQuery]] that binds it to some function that is able to produce a resulting value.
  */
class BoundLayerQuery[K, M, ReturnType](query: LayerQuery[K, M], f: LayerQuery[K, M] => ReturnType) {
  def where[F, T](params: LayerFilter.Expression[F, T])(implicit ev: LayerFilter[K, F, T, M]): BoundLayerQuery[K, M, ReturnType] =
    new BoundLayerQuery(query.where(params), f)

  def result: ReturnType = f(query)
}

File: geotrellis-python/geotrellis/spark/io/LayerQuery.py
from __future__ import absolute_import
from geotrellis.spark.KeyBounds import Bounds, EmptyBounds
from geotrellis.python.util.utils import flatten_list

class LayerQuery(object):
    def __init__(self, filterChain = lambda x: x):
        self._filterChain = filterChain

    def __call__(self, metadata):
        #bounds = metadata.getComponent(Bounds[K])
        bounds = metadata.getComponent(Bounds)
        if bounds is EmptyBounds:
            return []
        else:
            keyBounds = bounds
            _, keyBoundsList = self._filterChain((metadata, [keyBounds]))
            return keyBoundsList

    def where(self, exp, layerFilter):
        def newFilterChain(x):
            metadata, keyBoundsList = self._filterChain(x)
            filteredKeyBounds = [layerFilter(metadata, keyBound, exp) for keyBound in keyBoundsList]
            return (metadata, flatten_list(filteredKeyBounds))
        return LayerQuery(newFilterChain)

class BoundLayerQuery(object):
    def __init__(self, query, f):
        self.query = query
        self.f = f

    def where(self, params, layerFilter):
        return BoundLayerQuery(self.query.where(params), f)

    @property
    def result(self):
        return self.f(self.query)
